### ISSUE TITLE
Zero-dim Tensors (scalars) should be printed at full precision

### ISSUE THREAD

**Issue by colesbury** [pytorch/pytorch#21028]
## 🚀 Feature
The string representation of zero-dimensional Tensors should be sufficient to exactly reconstruct their value.

Ideally it should also print using the minimal number of digits. For example, we should avoid printing 1.0/10.0 as `0.100000001`.

## Motivation
We've had a number of bug reports where some value looks like 1.0 but is actually larger than 1. Many people don't know that you need to call `item()` to get an exact value.

NumPy uses the Dragon4 algorithm (Guy & Steele, 1990). Their implementation is here:
https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/dragon4.c

The algorithm is described here:
http://kurtstephens.com/files/p372-steele.pdf (retrospective, 2004)

**Comment by umanwizard** [comment]
Good idea

**Comment by umanwizard** [comment]
@colesbury is there any good reason we can't just use Python's built-in capability for printing numbers? (I.e., when printing a 0-dim tensor `t`, just print `'tensor(' + repr(t.item()) + ')'`

**Comment by colesbury** [comment]
@umanwizard, yeah you want things like 1.0 / 10.0 (in float32 Tensors) to print 0.1 not 0.10000000149011612.

**Comment by umanwizard** [comment]
For posterity: @colesbury 's comment above was a reference to mine (which I deleted) that said this:

> is there any good reason we can't just use Python's built-in capability for printing numbers? (I.e., when printing a 0-dim tensor t, just print 'tensor(' + repr(t.item()) + ')'

**Comment by umanwizard** [comment]
Not sure whether I will have time to port this dragon4 code to PyTorch, but I am going to try to.
